R/sortablestr2i.R

```diff
@@ -0,0 +1,42 @@
+# Alternative to hashstr2i that can produce semi-sorted chunks
+# Apply as e.g.:
+# split_values <- map(dff, sample_n, size=1) %>% 
+#   select(c("id1", "id2")) %>%
+#   collect() %>%
+#   arrange(id1, id2)
+# 
+# shard_by_rule <- splitstr2i(split_values)
+# code = glue::glue("df[,.out.disk.frame.id := {shard_by_rule}]")
+
+sortablestr2i <- function(split_values, desc_vars){
```

Comment:
I am a little concerned about the performance of this. I am not sure if it will slow, but it seems the fastest possible in R. I know `{data.table}` has `fifelse` but not sure how efficient that is for multi-statement `if`s.

Can we have some benchmarks that can be run? The `hash2stri` can compute billions of hashes every quickly.

Reply:
This was surprisingly not that much slower in my testing. 

Try this which benchmarks on 30x flights dataset (10m records):
 * hard_group_by process with "hash" vs the new "sort" splitter
 * the proposed disk.frame arrange vs dplyr arrange, and
 * the hash function vs the generated sort split function

```{r}
library(dplyr)
library(data.table)
library(disk.frame)
n_rep = 30
flights.df <- rbindlist(lapply(seq_len(n_rep), function(X) nycflights13::flights))
flights.dff <- as.disk.frame(flights.df)
flights.dt <- as.data.table(flights.df)

# Time to hard group by with hash:
tic <- Sys.time()
hgb1 <- flights.dff %>%
  hard_group_by(origin, year, month, day, shardby_function="hash")
(toc = Sys.time() - tic)
  
# Time to hard group by with sort:  
tic <- Sys.time()
hgb2 <- flights.dff %>%
  hard_group_by(origin, year, month, day, shardby_function="sort")
(toc = Sys.time() - tic)
  
# Time to sort with disk.frame:  
tic <- Sys.time()
harr <- flights.dff %>%
  hard_arrange(origin, year, month, day)
(toc = Sys.time() - tic)
  
# Time to Dplyr sort
tic <- Sys.time()
darr <- flights.df %>%
  arrange(origin, year, month, day)
(toc = Sys.time() - tic)

# Raw test of the functions - hashstr2i
hashstr2i <- function(x, ngrps) {
  .Call('_disk_frame_hashstr2i', PACKAGE = 'disk.frame', x, ngrps)
}

tic <- Sys.time()
flights.df[,.out.disk.frame.id := hashstr2i(paste0(day,month,origin,year), 4)]
(toc = Sys.time() - tic)

# Raw test of the functions - sortablestr2i
tic <- Sys.time()
flights.dt[,.out.disk.frame.id := ((origin > "EWR" | (origin == "EWR" & (year > 2013 | (year == 2013 & (month > 4 | (month == 4 & (day > 26 | (day == 26)))))))))* 1 + ((origin > "EWR" | (origin == "EWR" & (year > 2013 | (year == 2013 & (month > 9 | (month == 9 & (day > 19 | (day == 19)))))))))* 1 + ((origin > "JFK" | (origin == "JFK" & (year > 2013 | (year == 2013 & (month > 2 | (month == 2 & (day > 19 | (day == 19)))))))))* 1 + ((origin > "JFK" | (origin == "JFK" & (year > 2013 | (year == 2013 & (month > 7 | (month == 7 & (day > 17 | (day == 17)))))))))* 1 + ((origin > "JFK" | (origin == "JFK" & (year > 2013 | (year == 2013 & (month > 11 | (month == 11 & (day > 19 | (day == 19)))))))))* 1 + ((origin > "LGA" | (origin == "LGA" & (year > 2013 | (year == 2013 & (month > 4 | (month == 4 & (day > 4 | (day == 4)))))))))* 1 + ((origin > "LGA" | (origin == "LGA" & (year > 2013 | (year == 2013 & (month > 8 | (month == 8 & (day > 26 | (day == 26)))))))))* 1]
(toc = Sys.time() - tic)
```
Noting that I am on an older non-SSD Mac:
1. The "sort split" function is only about 50% slower than the hash2stri function.
2. This contributes a 10% slower time for hard_group_by overall
3. Overall this disk.frame hard_arrange is about four times slower than dplyr arrange. 

YMMV with an SSD.